tools/pd-api-bench/main.go

```diff
@@ -0,0 +1,225 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"context"
+	"crypto/tls"
+	"flag"
+	"fmt"
+	"log"
+	"net/http"
+	"os"
+	"os/signal"
+	"syscall"
+	"time"
+
+	pd "github.com/tikv/pd/client"
+	"github.com/tikv/pd/client/tlsutil"
+	"github.com/tikv/pd/pkg/utils/apiutil"
+	"github.com/tikv/pd/pkg/utils/typeutil"
+	"go.uber.org/zap"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/keepalive"
+)
+
+var (
+	pdAddr = flag.String("pd", "127.0.0.1:2379", "pd address")
+	// min-resolved-ts
+	minResolvedTSByGRPC = flag.Bool("min-resolved-ts-grpc", false, "min-resolved-ts by grpc qps")
+	minResolvedTSByHTTP = flag.Bool("min-resolved-ts-http", false, "min-resolved-ts by http qps")
+	// concurrency
+	client = flag.Int("client", 1, "client number")
+	// qps
+	qps = flag.Int("qps", 1000, "qps")
+	// tls
+	caPath   = flag.String("cacert", "", "path of file that contains list of trusted SSL CAs")
+	certPath = flag.String("cert", "", "path of file that contains X509 certificate in PEM format")
+	keyPath  = flag.String("key", "", "path of file that contains X509 key in PEM format")
+)
+
+var base = int(time.Second) / int(time.Millisecond)
+
+func main() {
+	flag.Parse()
+	ctx, cancel := context.WithCancel(context.Background())
+	sc := make(chan os.Signal, 1)
+	signal.Notify(sc,
+		syscall.SIGHUP,
+		syscall.SIGINT,
+		syscall.SIGTERM,
+		syscall.SIGQUIT)
+	var sig os.Signal
+	go func() {
+		sig = <-sc
+		cancel()
+	}()
+
+	if *minResolvedTSByGRPC {
+		go handleMinResolvedTSByGRPC(ctx)
+	}
+	if *minResolvedTSByHTTP {
+		go handleMinResolvedTSByHTTP(ctx)
+	}
+
+	<-ctx.Done()
+	switch sig {
+	case syscall.SIGTERM:
+		exit(0)
+	default:
+		exit(1)
+	}
+}
+
+func handleMinResolvedTSByGRPC(ctx context.Context) {
+	log.Println("handleMinResolvedTSByGRPC start...")
+	tt := base / *qps
+	for i := 0; i < *client; i++ {
+		pdCli := newPDClient()
+		go func() {
+			ticker := time.NewTicker(time.Millisecond * time.Duration(tt))
+			defer ticker.Stop()
+			for {
+				select {
+				case <-ticker.C:
+					_, _, err := pdCli.GetMinResolvedTS(ctx, []uint64{1, 2, 3})
+					if err != nil {
+						log.Println(err)
+						continue
+					}
+				case <-ctx.Done():
+					log.Println("Got signal to exit handleMinResolvedTSByGRPC")
+					return
+				}
+			}
+		}()
+	}
+}
+
+func handleMinResolvedTSByHTTP(ctx context.Context) {
+	log.Println("handleMinResolvedTSByHTTP start...")
+	// Judge whether with tls.
+	protocol := "http"
+	if len(*caPath) != 0 {
+		protocol = "https"
+	}
+	url := fmt.Sprintf("%s://%s/pd/api/v1/min-resolved-ts", protocol, *pdAddr)
```

Comment:
this tools call api bench, but seems it only bench `min-resolved-ts`

Reply:
This product is above 5706 and has been modified a lot. We are still working to support more API test expediently